### DISCUSSION TITLE
How to improve LLVM situation?

### DISCUSSION THREAD

**Discussion by abique** [microsoft/vcpkg#29391] [Q&A]
Hi,

I believe that today LLVM is in a poor situation within VCPKG:
- because of the "diamond" issue we can only have one version of LLVM within the ports
- it is very painful to update LLVM as it breaks the API on each update
- LLVM will soon be 2 versions behind

I've created https://discourse.llvm.org/t/llvm-in-vcpkg-and-painful-update/68051 to discuss possible improvements.
It'd be amazing to have a VCPKG maintainer's point of view on this.

I believe the right path is to:
- have a C++ `LLVM_NAMESPACE` so llvm-17 can be in the namespace `llvm17` and avoid clashing with llvm-14
- we then would be able to have multiple versions of LLVM within the ports

What do you think?

**Comment by JackBoosY** [comment]
cc @BillyONeal

**Comment by BillyONeal** [comment]
I'm not sure we necessarily want multiple versions of llvm in the catalog. As was correctly pointed out in the linked discussion, that still makes anything that needs to touch multiple versions extremely painful. IMO this is just one of the consequences of LLVM breaking stuff every 6 months making LLVM a painful dependency to take.

Even if the suggested namespace solution happens, it would still make any resulting program need to bundle multiple copies of LLVM, which, at ~60MB a pop, is.... a lot.

As for our copy being versions behind, there is some precedent for leaving things behind if absolutely necessary, as we did with `tbb` leaving behind `usd`, but that's a judgement call / balancing test we would have to make when it happens.

**Reply by abique** [reply]
As a user, I'd appreciate to be able to enjoy the latest version of LLVM as well as I'd understand that some libraries may depend on older version.

Would it be an option to allow multiple versions of LLVM to coexist within VCPKG and simply prevent the diamond issue in a soft way: forbid `./vcpkg install <port>` if the given port is conflicting with an already installed package? Or even simpler, just let the build fail at link time.

Because as time goes on, I imagine that it is harder and harder to upgrade LLVM because of the increasing number of ports depending on it. It leads to an "abnormal?" pressure being put on anyone trying to upgrade LLVM.

**Reply by abique** [reply]
Also, updating LLVM is such a major task that among all the people who did it so far, no one wants to do it again, and somebody tried for version 15 and gave up.

**Reply by BillyONeal** [reply]
>Would it be an option to allow multiple versions of LLVM to coexist within VCPKG and simply prevent the diamond issue in a soft way: forbid ./vcpkg install <port> if the given port is conflicting with an already installed package?

We allowed edge cases like this for libressl and boringssl (as replacements in the registry for openssl) but stopped accepting this type of thing when registries became a thing. The problem with adding new such replacements is that we do not have the capability to verify that they work (both for 'don't have the feature' and 'quickly becomes impossible in terms of machine time to test the combinatoric explosion it creates' reasons).

>Also, updating LLVM is such a major task that among all the people who did it so far, no one wants to do it again, and somebody tried for version 15 and gave up.

We can make the call that having up to date is more important than having things depending upon it in the curated registry, but it isn't practical to allow multiple versions at once.

**Reply by abique** [reply]
> > Would it be an option to allow multiple versions of LLVM to coexist within VCPKG and simply prevent the diamond issue in a soft way: forbid ./vcpkg install  if the given port is conflicting with an already installed package?
> 
> We allowed edge cases like this for libressl and boringssl (as replacements in the registry for openssl) but stopped accepting this type of thing when registries became a thing. The problem with adding new such replacements is that we do not have the capability to verify that they work (both for 'don't have the feature' and 'quickly becomes impossible in terms of machine time to test the combinatoric explosion it creates' reasons).

Understood.

> > Also, updating LLVM is such a major task that among all the people who did it so far, no one wants to do it again, and somebody tried for version 15 and gave up.
> 
> We can make the call that having up to date is more important than having things depending upon it in the curated registry, but it isn't practical to allow multiple versions at once.

I'd be in favor of that as I'm a direct user of LLVM for JIT compilation. But then I'll be in conflict with users of libraries depending on LLVM which do not compile or hasn't been tested with the latest LLVM version yet. And I don't how you'll cut but you won't be able to make every one happy.

Even if you allow 2 or 3 months for libraries to upgrade and support the latest LLVM, it may not work with the libraries agenda, and will you be able to drop those libraries?

I still believe that the only peaceful solution is to allow multiple versions to coexists, ideally with support from LLVM for `LLVM_NAMESPACE` to avoid symbol clash.

**Reply by BillyONeal** [reply]
@dan-shaw @ras0219-msft @vicroms @markle11m @robotodad @JavierMatosD and I discussed this today.

1. We agree it remains that there can only be one LLVM version selected by default in the curated registry.
2. We seem to be in support of providing ways for customers to get a different version of LLVM, which prevent them from using anything else in the curated registry that needs LLVM. We do need some technical solutions, in particular:
    * LLVM takes *forever* to build, so we will probably need to be talking about getting budget to do that.
    * We need to fix the binary caching issue where we can't upload cache entries of LLVM before going here.
    * The alternatives proposal is probably relevant here.
3. We agree that we need some sort of policy for things that depend on LLVM in vcpkg. For example, "vcpkg will be on LLVM N-1 and you must be ready to adopt that when LLVM N comes out. For example, when LLVM 16 comes out we will change our canonical version to LLVM 15, and if you can't adopt that at that time then you are delisted from the curated registry." We did not reach consensus on specifics or details of this policy.

**Reply by abique** [reply]
Hi @BillyONeal,
Thank you very much for taking the time to think this through.
Your conclusions seems the right ones to me, and are independent of a possible `LLVM_NAMESPACE`.

Will your team proceed with this plan?

**Reply by BillyONeal** [reply]
>Will your team proceed with this plan?

I'm sorry, perhaps I was unclear, there isn't really a plan enumerated above, only a list of points on which we achieved consensus.

**Reply by Neumann-A** [reply]
LLVM 16 has been released. So who is going to update llvm to 15? MSVC headers already require 15.
(Maybe somebody already has a working overlay?)

**Reply by abique** [reply]
There was a discontinued PR for LLVM 15, which I'm using.

**Reply by Neumann-A** [reply]
Hmm this one: #26902?

**Reply by abique** [reply]
Yes, I took it and bumped to 15.0.7.

**Reply by abique** [reply]
If anyone has an overlay for 16 I'm interested :)

**Reply by Neumann-A** [reply]
If halide and mesa get bumped in vcpkg, vcpkg could have 15.0.7 easily since both claim support for llvm 15
